### ISSUE TITLE
Slashes in tag gives a repository error instead of a tag error

### ISSUE THREAD

**Issue by baileysage** [google/go-containerregistry#762]
We were using $CI_COMMIT_REF_NAME in a CI pipeline as our tag and someone used '/' in their branch name. 

The error we got was: 
`repository can only contain the runes `abcdefghijklmnopqrstuvwxyz0123456789_-./`: /registry.exampleurl.com/private/repo-name:test/withslash`

We've corrected our pipeline, but it would be nice to see the error happen on the forbidden character in the tag rather than failing on the repository check.

**Comment by jonjohnsonjr** [comment]
Yeah... the `name` package leaves a lot to be desired. I'd like to redo basically everything so that we can produce better error messages.
